### ISSUE TITLE
Load config.ru with relative path when possible, and change how current working directory is handled

### ISSUE THREAD

**Issue by felixbuenemann** [phusion/passenger#1596]
After upgrading from passenger-nginx 5.0.15 to 5.0.16 our ruby 2.1.6 / spree commerce 1.1 / rails 3.2.22 app ist crashing with `SystemStackError: stack level too deep` when doing a POST request to put items in the cart. After downgrading to passenger 5.0.15 the app works fine again, so this seems to be a regression.

The exception started shortly after upgrading from 5.0.15 to 5.0.16 with no other changes to the system.

**Comment by felixbuenemann** [comment]
The backtrace is not very helpful in this case and only contains one line, in this case the newrelic_rpm rack middleware. But if we remove it it crashes in the next middleware layer. So my guess is that passenger is sending a malformed rack request.

The app is using rack 1.4.6. I'll try reproducing in development environment using passenger standalone.

**Comment by OnixGH** [comment]
Thanks for reporting, that would be most helpful.

Stack level overflow sounds like runaway recursion / redirection, are you seeing any signs of this in the logs? Is your app deployed on a sub-uri?

**Comment by felixbuenemann** [comment]
No, it's nunning in the docroot, pretty much default passenger config. OS is Ubuntu 12.04.5 LTS x86_64 and ruby is from the brightbox ruby ng repo.

**Comment by felixbuenemann** [comment]
I haven't yet been able to reproduce with passenger standalone. But the new `seek(*args)` method added to the tee input helper in 5.0.16 is recursive, so it could potentially cause a system stack error if it never aborts.

**Comment by felixbuenemann** [comment]
Sorry, it's called on @tmp os it's not recursive. I'll try upgrading back to 5.0.16 after business hours and see if that's the code causing the issue.

**Comment by felixbuenemann** [comment]
Tried upgrading ruby to 2.1.7, rack to 1.4.7 and back to passenger 5.0.16, still crashes while 5.0.15 is fine. Tried uncommenting `seek` from TeeInput and restart nginx on 5.0.16, no change.

What would be a good way to debug what changed in the request between 5.0.15 and 5.0.16?

**Comment by felixbuenemann** [comment]
OK, after upgrading the app to ruby 2.2 to get a useful stacktrace I figured out the problem.
Spree's idea of Decorators is using class_eval on existing classes to allow customization inside the app. One decorated controller used `alias_method_chain` to inject extra authorization. For some reason passenger 5.0.16 seems to load that file twice, so the alias_method_chain creates an endless loop.

I worked around the problem by checking if the method is already defined before aliasing it (I didn't dare introducing `Module#prepend` into such an ancient codebase).

It's entirely unclear to me however why this is caused at all and why the behavior changed in passenger 5.0.16, because the code for loading decorators in `config/application.rb` looks likes this:
```ruby
module MyApp
  class Application < Rails::Application
    config.to_prepare do
      # Load application's model / class decorators
      Dir.glob(File.join(File.dirname(__FILE__), "../app/**/*_decorator*.rb")) do |c|
        Rails.configuration.cache_classes ? require(c) : load(c)
      end

      # Load application's view overrides
      Dir.glob(File.join(File.dirname(__FILE__), "../app/overrides/*.rb")) do |c|
        Rails.configuration.cache_classes ? require(c) : load(c)
      end
    end
end
```

Both the staging and production environment have cache_classes enabled, so they should never load the code more than once.

**Comment by FooBarWidget** [comment]
You can see what has changed between 5.0.15 and 5.0.16 by running this in the Passenger repo:

    git diff release-5.0.15..release-5.0.16

I don't see any change which may influence this though. In any case, Passenger shouldn't have any influence on how code is loaded; all it does is loading config.ru.

Can you debug this by putting a 'puts caller' in the file that is loaded twice? That way you can see which functions are responsible for loading that file.

**Comment by felixbuenemann** [comment]
@FooBarWidget Yea, I already looked at the code and none of the changes looked like it could cause the behavior, but the problem was 100% reproducible on the staging/production server.

I'll try the `puts caller` trick later.

**Comment by felixbuenemann** [comment]
On 5.0.16 the file gets loaded twice by `/usr/share/passenger-enterprise/helper-scripts/rack-preloader.rb`, but only once on 5.0.15.

I diffed the rack-preloader.rb between 5.0.15 and 5.0.16 and they are identical, but it could just be the code that's triggering the preloader.

**Comment by dmitry** [comment]
Can you list what have been changed between passenger releases? Or you are trying only to change passengers, and on 5.0.16 the code fails to start because of double execution?

**Comment by felixbuenemann** [comment]
@dmitry I'm not sure what you're trying to say.

The preloader gets triggered twice by the same Passenger AppPreloader process id on 5.0.16 and only once on 5.0.15.

The calltree looks identical:
```
…
/var/www/myapp/web/staging/current/config.ru:1:in `new'
/var/www/myapp/web/staging/current/config.ru:1:in `<main>'
/usr/share/passenger-enterprise/helper-scripts/rack-preloader.rb:91:in `eval'
/usr/share/passenger-enterprise/helper-scripts/rack-preloader.rb:91:in `preload_app'
/usr/share/passenger-enterprise/helper-scripts/rack-preloader.rb:137:in `<module:App>'
/usr/share/passenger-enterprise/helper-scripts/rack-preloader.rb:13:in `<module:PhusionPassenger>'
/usr/share/passenger-enterprise/helper-scripts/rack-preloader.rb:12:in `<main>'
```

To test I downgrade passenger to 5.0.15 using this script:
```sh
apt-get install passenger-enterprise=1:5.0.15-2~precise1 passenger-enterprise-dev=1:5.0.15-2~precise1 nginx-common=1:1.8.0-9.5.0.15~precise1 nginx-extras=1:1.8.0-9.5.0.15~precise1
```
And to get back to 5.0.16 I simply do a `apt-get upgrade`.

**Comment by dmitry** [comment]
@felixbuenemann have you tried to reproduce an error with newly generated rails application?

**Comment by felixbuenemann** [comment]
@dmitry No, I can't event reproduce it with the same app using passenger-enterprise standalone on my local mac.

**Comment by felixbuenemann** [comment]
This seems to be related to symlinks. The preloader is only loading the config.ru once, but I can see that the decorator file has a different `__FILE__`between the two requires:
```
/var/www/myapp/web/staging/current/app/controllers/orders_controller_decorator.rb
/var/www/myapp/web/staging/releases/20150901211622/app/controllers/orders_controller_decorator.rb
```
So it gets required once with the symlinked release path and once with the expanded release path.

This bug is getting weirder the longer I look at it :smile:

**Comment by felixbuenemann** [comment]
So there are indeed changes in how the app preloader is called. The options passed in the handshake differ:
```diff
--- log_5.0.15.txt	2015-09-02 19:51:01.000000000 +0200
+++ log_5.0.16.txt	2015-09-02 19:47:41.000000000 +0200
@@ -1,16 +1,15 @@
 passenger_root: /usr/lib/ruby/vendor_ruby/phusion_passenger/locations.ini
-passenger_version: 5.0.15
 ruby_libdir: /usr/lib/ruby/vendor_ruby
-gupid: 16e851e-9IRp2VBpPe
+passenger_version: 5.0.16
 UNIX_PATH_MAX: 108
 app_root: /var/www/insys/web/staging/current
 app_group_name: /var/www/insys/web/staging/current/public (staging)
 app_type: rack
 start_command: /usr/bin/ruby	/usr/share/passenger-enterprise/helper-scripts/rack-loader.rb
-startup_file: config.ru
+startup_file: /var/www/insys/web/staging/current/config.ru
 process_title: Passenger RubyApp
 log_level: 3
 start_timeout: 90000
```

**Comment by felixbuenemann** [comment]
To come full circle on the problem:

The code loader snippet from `config/application.rb` [posted above](https://github.com/phusion/passenger/issues/1596#issuecomment-136868183) is also used by various rails engines used by spree commerce plugins, which are part of the main spree commerce bundle. The initializers from one of these engines then seems to use the expanded paths causing the double require.

Given that the code posted is used by all current spree version, potentially all versions of spree commerce deployed with capistrano are potentially affected by this problem. Arguably the right fix would be to change all those code loader snippets to use `File.expand_path`, but I'm not sure thats viable.

**Comment by FooBarWidget** [comment]
Wow, good debugging. I never thought that something as innocent-looking as absolutizing the config.ru path could break application loading.

I've committed a fix in the `GH-1596` branch. Could you give it a try?

**Comment by felixbuenemann** [comment]
Thanks, I'll try the fix this evening.

**Comment by felixbuenemann** [comment]
I'm not sure the fix will work, because Dir.pwd in ruby resolves symlinks:
```sh
$ pwd
/var/www/myapp/web/production/current
ruby -e "puts Dir.pwd"
/var/www/myapp/web/production/releases/20150903172004
ruby -e "puts File.dirname('/var/www/myapp/web/production/current/config.ru') == Dir.pwd"
false
```
Couldn't you compare to `options["app_root"]` instead?

**Comment by felixbuenemann** [comment]
Hmm, ruby 1.8.7 doesn't habe `File.realname`, only `File.readlink`…

**Comment by FooBarWidget** [comment]
Urgh, totally forgot about that. I updated the branch. How about now?

**Comment by felixbuenemann** [comment]
Unfortunately it still uses the resolved path, maybe because passenger already resolves the symlink before cding into the app dir?
```
App 9390 stderr: YYY `pwd` => /var/www/insys/web/staging/releases/20150904125955
App 9390 stderr: YYY File.dirname(/var/www/myapp/web/staging/current/config.ru) => /var/www/myapp/web/staging/current
App 9390 stderr: YYY Dir.pwd_no_resolve => /var/www/myapp/web/staging/releases/20150904125955
```

**Comment by FooBarWidget** [comment]
I can't reproduce that. `Pwd.pwd_no_resolve` properly prints the original symlink path for me, not the resolved path. This suggests that you may be affected by another issue.

Where did you insert those debugging calls?

**Comment by felixbuenemann** [comment]
I added the debug code directly in the rack-preloader where it calls the new maybe… helper. I don't think that either rails 3.2 or spree do a chdir, but I'd have to grep through the source tree to make sure. 

Did you test `Dir.pwd_no_resolve` inside the rack preloader?

**Comment by FooBarWidget** [comment]
Yes I did.

What happens if you call `puts Dir.pwd_no_resolve` at the very beginning of the `preload_app` method? Does that yield any different results?

Can you also open ext/common/SpawningKit/Spawner.h and look for this line?

    ret = chdir(info.appRootPathsInsideChroot.back().c_str());

Before that line, add:

    P_WARN("C++ chdir: " << info.appRootPathsInsideChroot.back());

Then run 'rake nginx' inside the Passenger source directory. What do the logs tell you when you start your app now?

**Comment by felixbuenemann** [comment]
I'm not near a computer right now, but I can check later. Did you test nginx with user switching? (App user != nginx user)

**Comment by felixbuenemann** [comment]
At the first line of preload_app in rack-preloader.rb I added the following code:
```ruby
def self.preload_app
  STDERR.puts "XXX Dir.pwd_no_resolve=#{Dir.pwd_no_resolve} `pwd`=#{`pwd`.strip}"
  # …
end
```
Output:
```
App 57150 stderr: [ 2015-09-06 22:23:20.2747 57150/7fa704159700 Spa/Spawner.h:980 ]: C++ chdir: /var/www/myapp/web/staging/current
App 57150 stderr: XXX Dir.pwd_no_resolve=/var/www/myapp/web/staging/releases/20150904125955 `pwd`=/var/www/myapp/web/staging/releases/20150904125955
```

**Comment by felixbuenemann** [comment]
I think the reason is that the symlink is resolved by chdir in certain situations.

I wrote the following small c program:
```c
#include <stdio.h>
#include <unistd.h>
#include <stdlib.h>
#include <errno.h>

const char * const path = "/var/www/insys/web/staging/current";

int main() {
   printf("Changing directory to %s\n", path);

   if (chdir (path) == -1) {
       printf("chdir failed - %d\n", strerror (errno));
   } else {
       system("pwd");
       printf("\n");
   }
   return 0;
}
```
The output of pwd now depends on the source directory from where the code is called:
```sh
$ pwd
/var/www/myapp/web/staging/current
$  /tmp/chdir-test
Changing directory to /var/www/insys/web/staging/current
/var/www/insys/web/staging/current
$ cd
$ pwd
/var/www/myapp
$ /tmp/chdir-test
Changing directory to /var/www/myapp/web/staging/current
/var/www/myapp/web/staging/releases/20150904125955
```
You might want to look at `man 7 path_resolution`on linux, to see how `chdir` resolves the path:
> If the component is found and is a symbolic link (symlink), **we first resolve this symbolic link** (with the current lookup directory as starting lookup directory). Upon error, that error is returned. If the result is not a directory, an ENOTDIR error is returned. If the resolution of the symlink is successful and returns a directory, we set the current lookup directory to that directory, and go to the next component. Note that the resolution process here involves recursion. In order to protect the kernel against stack overflow, and also to protect against denial of service, there are limits on the maximum recursion depth, and on the maximum number of symbolic links followed. An ELOOP error is returned when the maximum is exceeded ("Too many levels of symbolic links").

**Comment by FooBarWidget** [comment]
Wow, that is surprising. No wonder I couldn't reproduce it, I was on OS X. Still, I didn't expect this in Linux either, because when working in bash I never saw this symlink resolution behavior:

    $ cd /var/www/myapp/current
    $ pwd
    /var/www/myapp/current

After some investigations, it turns out that `pwd` is a *shell builtin*. If I use /bin/pwd then it prints the symlink:

    $ /bin/pwd
    /var/www/myapp/web/staging/releases/20150904125955

I've updated the the branch, everything should now work. Can you confirm? I'm hoping to release the fix this week.

**Comment by felixbuenemann** [comment]
Btw. chdir behavior on OS X is identical (same chdir.c as above, slightly tweaked to silence warnings):
```sh
$ ls -al /Users/felix/dev/sites/myapp/tmp/foo
lrwxr-xr-x  1 felix  staff  2  7 Sep 04:09 /Users/felix/dev/sites/myapp/tmp/foo -> ..
$ cd ~/tmp
$ ./chdir-test
Changing directory to /Users/felix/dev/sites/myapp/tmp/foo
/Users/felix/dev/sites/myapp
$ cd /Users/felix/dev/sites/myapp/tmp/foo
$ ~/tmp/chdir-test
Changing directory to /Users/felix/dev/sites/myapp/tmp/foo
/Users/felix/dev/sites/myapp/tmp/foo
```

**Comment by FooBarWidget** [comment]
I'm getting very weird result in OS X. I have the following directory structure:

    ~/current -> releases/x
    ~/releases/
         x/

I've written the following Python script. As far as I know, the calls in Python map directly to system calls.

~~~python
import os

os.chdir("/Users/hongli/current")

print("getcwd():")
print(os.getcwd())

print("")
print("/bin/pwd:")
os.system("/bin/pwd")
~~~

If I run this from /users/hongli, I get:

~~~
getcwd():
/Users/hongli/releases/x

/bin/pwd:
/Users/hongli/releases/x
~~~

...which matches your observations. But if I first use the shell to chdir to /Users/hongli/current, and then I run the script, then I get:

~~~
getcwd():
/Users/hongli/releases/x

/bin/pwd:
/Users/hongli/current
~~~

So on OS X, pwd is doing something different but I have not yet figured out what.

Symlink resolution rules are very strange indeed...

**Comment by FooBarWidget** [comment]
After reading http://www.opensource.apple.com/source/shell_cmds/shell_cmds-170/pwd/pwd.c, it seems that `pwd` on OS X gets the logical working directory from bash, through the PWD environment variable. But inside Ruby we can't count on that being accurate.

**Comment by FooBarWidget** [comment]
I realized just now that it's actually a bit dangerous to use the `pwd` command because both the GNU and the BSD version read from the `PWD` environment variable, which may not be correct. I guess I need to rethink how to handle paths and symlinks...

**Comment by felixbuenemann** [comment]
Do we need to deal with pwd at all? In my experience the problem didn't occur, as along as I used the fully expanded path, so if we always expand the config.ru path we should be fine.

**Comment by FooBarWidget** [comment]
A new fix has been committed that takes the issues with working directories and symlinks into account. Could you try the GH-1596 branch again? The branch has been rebased so you will need to 'git fetch && git reset --hard origin/GH-1596`.

We need to deal with pwd because the application may change the working directory before we load config.ru.

**Comment by felixbuenemann** [comment]
@FooBarWidget The normal case where only your nging `root` contains symlinks is fixed. But if your `passenger_startup_file` contains symlinks, the problem still occurs.

Example:
```
root /apps/foo/current;
passenger_startup_file symlinked_sub_app/config.ru;
```

**Comment by felixbuenemann** [comment]
Btw. I think I'm going to propose a change to spree to work around the issue, which is pretty simple. Unfortunately that won't help with all the existing spree installations, unless they are updated.

**Comment by FooBarWidget** [comment]
Does Spree use the `symlinked_sub_app/config.ru` format?

**Comment by felixbuenemann** [comment]
No, it's just something I noticed while testing locally, so I thought I'd mention it.
